[anneal] Add more thorough agent documentation (#3489)
google/zerocopy: anneal/agent_docs/design/principles.md

- arbitrary Rust codebases and use cases.

These are architectural scope constraints, not claims about the current
executable or requirements that the first milestone support every Rust
construct and property domain.

Unsafe Rust creates proof obligations that Rust's type system does not
discharge. Giving those obligations a rigorous, practical treatment is a
primary Anneal use case. This demands unusually strong support for memory
safety in unsafe-heavy systems code, but Anneal is not limited to memory
safety. Functional correctness, protocol conformance, panic freedom, resource
bounds, and user-defined properties must be able to participate in the same
verification framework.

Anneal is intended for ordinary Rust organizations, not only
formal-methods specialists. Specialists and AI agents may help author proofs
and evolve specifications, but Rust engineers must be able to understand,
debug, review, and incrementally adopt the resulting workflow.

## Preserve soundness and claim integrity

Rust compilers reserve the right to give arbitrary behavior to a program after
undefined behavior. Anneal's source models likewise promise correspondence

> only while Rust's validity requirements hold. Soundness is therefore not just

Reviewer: ```suggestion
only when Rust program's behavior is defined. Soundness is therefore not just
```

one optional property among peers: it is a prerequisite for interpreting a
model theorem as a fact about the Rust artifact.

Anneal must account for every operation whose validity is required by the
supported Rust semantics and enforce adequate obligations for it. A weak,
incomplete, or vacuous user specification must not be able to redefine a Rust
soundness requirement away. Unsupported semantics, absent coverage, and
residual trust must qualify or prevent the reported claim rather than
disappear behind a successful proof check.

User-defined properties have a different source of meaning. Anneal must
faithfully enforce a user's definition of cryptographic correctness, protocol
conformance, or another domain property, but cannot infer whether that
definition captures the user's intent. Whenever a specification claims
correspondence with Rust or another external semantics, however, its adequacy
is an auditable trust question rather than a matter of personal intent.

## Prefer local, compositional reasoning

Unsafe code can implement a safe API only when the implementation can be
checked against an abstraction boundary that every supported safe client may

> then rely on. Clients should consume established interface guarantees rather

Reviewer: ```suggestion
checked against an abstraction boundary that the implementer has proven sound.
Clients should consume established interface guarantees rather
```

than reopen private implementations or reason about an unrelated remainder of

> the program.

Reviewer: I wonder whether we should have a stronger statement that refers to [non-interference](https://en.wikipedia.org/wiki/Non-interference_(security)) here (between a lexically scoped abstraction and the rest of the program).

The durable formulation is contextual refinement: an implementation realizes
its declared interface in every supported context. The interface may include
values, effects, capabilities, resources, protocols, and invariants. Locality
does not require pretending that every implementation is a pure value
transformer.

Composition then has a uniform shape. Each item establishes its guarantees
under its assumptions, and each use establishes the requirements of the
operations it invokes. Property dependencies are evidence-graph edges. For
example, a callee's ordinary functional guarantee that an index is below a
buffer length may discharge the soundness requirement of a caller's raw
pointer access.

> 

Reviewer: There's something about how this is written that fails to provide clarity for me. I think we should cut these two paragraphs and draft a brief statement about composition to wrap up the paragraph above.

## Prefer the simplest faithful abstraction

Simple mathematical interfaces improve proof readability, stability, and
automation. Prefer a pure functional contract when it faithfully captures the
abstraction boundary. Preserve resource, provenance, initialization,
ownership, concurrency, protocol, and effect semantics whenever erasing them

> could invalidate the claim.

Reviewer: Suggestion: replace "Preserve" with "Introduce" and "erasing" with "eliding" or "ignoring".

This rule leads to a hybrid model rather than making “hybrid” an end in itself.
Some Rust can use Aeneas's simple functional interpretation, while other
operations require richer semantics. The exact supported boundary changes as
Aeneas evolves; consult the revision-sensitive
[Aeneas and Charon reference](../reference/aeneas-and-charon.md) rather than

> maintaining a second caveat list here.

Reviewer: This reads to me too much like "directions for an agent" when this documentation should be useful and meaningful to a human reader.

```suggestion
This principle suggests to a hybrid model, but does not make “hybrid” an end in itself.
Since the facets of rust semantics supported by Aeneas and Charon will evolve over
time, please refer to the
[Aeneas and Charon reference](../reference/aeneas-and-charon.md) for details
regarding how various semantic domains are managed by Anneal and its
dependencies.
```

Lifting a resource assertion into Lean must not turn it into an unrestricted
fact. Its rules for duplication, consumption, framing, opening, and
re-establishment remain part of its meaning. Separation logic is one important
way to preserve such disciplines, not the only permitted mechanism.

> 

Reviewer: I don't think this says enough precisely enough to be valuable. Either it needs to be significantly tightened up and given an example, or we should cut it.

be narrow, but their abstractions should be exercised against concrete Rust
programs and should leave room to extend coverage without replacing the core
composition model.

Lean, Aeneas, and their libraries already provide proof, weakest-precondition,
simplification, and tactic infrastructure. Build on maintained abstractions
when they meet Anneal's requirements. Replacing them is permissible when a
concrete benefit justifies the additional semantics and maintenance.
Likewise, share common proof reasoning where possible without presupposing
whether every property must use one symbolic execution or one proof backend.

Prefer robust programmatic integration points over textual patching or an
unreconciled source-level shadow of compiler-resolved item identities and
metadata. This does not decide whether Anneal should maintain a safety-tracking
system alongside Rust's own `unsafe` machinery. Changes to Aeneas and Charon
are in scope, as are longer-term changes to Rust and its specification.
Upstream and downstream ownership remain case-specific decisions that must
account for the burden placed on collaborators.

> 

Reviewer: I find this whole section hard to follow. Can we give a much crisper, shorter articulation of a principle, and use one or two simple examples to illustrate? I think that would be a lot clearer. Is there an example of the general form, "we could add bespoke first-class support for [narrow-rust-pattern], but Aeneas already supports [abstract-pattern]. By this principle, we support [narrow-rust-pattern], then, as a special case of [abstract-pattern] as follows: [some Lean and/or rust code]".

## Make trust explicit, classified, and reducible

> 

Reviewer: We should explain "what trust means" to Anneal. I _think_ we want something in the neighbourhood of "abstractions with semantics admitted without proof", but trust can mean different things in different contexts, even in security. A crisp statement of _what we mean by trust_ will lend clarity to this subsection.

It may be that we need multiple notions of trust, too. What it means to trust a compiler or other tool might be different from trust the way tools composed, which might be different from trusting things spelled `axiom` in a Lean model. (The last of this is what I had in mind with my vague definition above.)

unsupported coverage, and tool failures support different conclusions and
must remain distinguishable.

Trust must also be capable of shrinking. A formal Rust, library, compiler,
instruction-set, foreign-library, or hardware model should be able to replace
a shallower assumption without forcing every client contract to change.
Moving an assumption into a helper whose body is not modeled, or into an
upstream component, does not
reduce trust unless the end-to-end dependency has actually been removed.

## Support honest incremental adoption by ordinary Rust engineers

Real Rust codebases cannot usually formalize every obligation at once. Anneal
must support incremental adoption, including some form of existing prose
`// SAFETY:` justification in place of selected formal proofs. It may also
need a first-class representation of a proof that has not yet been completed.

Incremental adoption must not turn an assumption into a proof. Reports and
diagnostics must make clear what was checked, what remains, what claim that
evidence supports, and what would strengthen it. Favor stable proof interfaces,
diagnostics tied to Rust source, resilience to small program changes, and

> failures that explain the outstanding obligation.

Reviewer: This is another example of "written as an agent instruction" inside documentation that should be approachable and pleasant for humans to read. (I admit there is some tension here between my request to "state an actual principle"--which will often sound directive--and "write documentation, not pithy agent instructions".) Maybe even "This principle favors" or similar would make this feel more like documentation that agent instructions.

AI assistance may reduce the amount of formal expertise needed to author a
proof. It may not excuse an interface whose meaning or trust boundary Rust
engineers cannot audit. Performance and automation matter because feedback
latency shapes adoption, but neither may weaken or obscure the reported claim.

> 

Reviewer: I think we should either drop this or give it its own principle.

In terms of style, this paragraph feels like its talking around what it means to say, rather than being straightforward.

I _think_ the intent here is "don't let agents weasel their way out of required proof obligations". This aspiration feels like an implementation detail around syntax and tooling affordances that we are a long way from being able to address. I think the core principle here is mostly covered by "Make trust explicit, ...". It should be hard for humans or agents to "work around" any "hard obligations" that should be "difficult to erase". I see this as a consequence of "Make trust explicit, ...", but if we want to make it its own thing, I don't think it should be a paragraph under incremental adoption.

## Applying the principles

The principles often reinforce one another, but they do not define a permanent
total order. When they pull in different directions:

1. state the exact claim each alternative would support;
2. identify added trust, lost coverage, or erased semantic information;
3. ask whether the result still composes at an abstraction boundary;
4. consider usability, maintenance, and the path to broader coverage;
5. prefer reversible experiments while evidence is weak; and
6. record an irreversible choice explicitly.

A convenient design that compromises soundness or claim integrity is
unacceptable. Among designs that preserve those constraints, the balance of
simplicity, coverage, maintenance, and user experience is a case-specific

> judgment.

Reviewer: I think humans engaged in human dialogue are going to decide together how to apply these principles. I would advocate for moving this to `### Instructions for agents applying these principles` under `## Goals and scope`.
